### PR TITLE
[WFLY-10632] RESTEasy 3.6.0

### PR DESCRIPTION
**Pull request by asoldano**
This is a preliminary PR for the upgrade to RESTEasy 3.6 series (https://issues.jboss.org/browse/WFLY-10632).
It also includes the needed changes for RESTEASY-1911 and for removing the ee8.preview.mode switch for JAX-RS / JSON-B.

The idea would be to test / review this here and if / when we're fine with the changes here, we tag 3.6.0.Final, update the PR and merge it.

*Update - 29th June 2018*
The PR is now including RESTEasy 3.6.0.Final artifacts, so it's ready for being merged.

### CODE REVIEW

**Comment by asoldano** [comment]
retest this please (I opened the PR a little bit too early, while the artifacts were still being deployed to the repo, sorry)

**Comment by asoldano** [comment]
@jamezp @kabir @rsvoboda I believe this is ready for review. Can you please have a look or tag whoever can do that here? Once I have a "go", I can tag/release RESTEasy 3.6.0.Final and update the PR so that it can be really merged.

**Comment by asoldano** [comment]
Adding @marekkopecky and @kanovotn

**Review by kabir — CHANGES_REQUESTED** [review]
I think you need full-feature-pack-licenses.xml for the two new dependencies. Apart from that it looks fine from my POV

**Comment by asoldano** [comment]
@kabir , thanks, added

**Review by kabir — APPROVED** [review]

File: galleon-pack/pom.xml [hunk 1]
```diff
@@ -1330,12 +1330,12 @@
...
             <groupId>org.jboss.spec.javax.ws.rs</groupId>
-            <artifactId>jboss-jaxrs-api_2.0_spec</artifactId>
+            <artifactId>jboss-jaxrs-api_2.1_spec</artifactId>
         </dependency>
 
         <dependency>
-            <groupId>org.jboss.spec.javax.ws.rs</groupId>
-            <artifactId>jboss-jaxrs-api_2.1_spec</artifactId>
+            <groupId>javax.json.bind</groupId>
+            <artifactId>javax.json.bind-api</artifactId>
```
Comment (jamezp) [inline galleon-pack/pom.xml]: Is this required here? It should be coming in from the `servlet-galleon-pack`.

File: servlet-feature-pack/src/main/resources/modules/system/layers/base/javax/json/bind/api/main/module.xml [hunk 2]
```diff
@@ -22,11 +22,7 @@
         <property name="jboss.api" value="public"/>
     </properties>
     <resources>
-        <artifact name="${javax.json.bind:javax.json.bind-api}">
-            <conditions>
-                <property-equal name="ee8.preview.mode" value="true"/>
-            </conditions>
-        </artifact>
+        <artifact name="${javax.json.bind:javax.json.bind-api}" />
```
Comment (jamezp) [inline servlet-feature-pack/src/main/resources/modules/system/layers/base/javax/json/bind/api/main/module.xml]: This is going to conflict with #11370. I'd rather see either #11370 merged first or this in a separate commit.

File: pom.xml [hunk 3]
```diff
@@ -324,7 +324,7 @@
         <version.org.jboss.mod_cluster>1.3.9.Final</version.org.jboss.mod_cluster>
         <version.org.jboss.narayana>5.8.1.Final</version.org.jboss.narayana>
         <version.org.jboss.openjdk-orb>8.1.1.Final</version.org.jboss.openjdk-orb>
-        <version.org.jboss.resteasy>3.5.1.Final</version.org.jboss.resteasy>
+        <version.org.jboss.resteasy>3.6.0.CR1</version.org.jboss.resteasy>
```
Comment (jamezp) [inline pom.xml]: We probably need some clarification about the rules, but we're supposed to be always final for dependencies.

**Review by jamezp — CHANGES_REQUESTED** [review]

File: servlet-feature-pack/src/main/resources/modules/system/layers/base/javax/json/bind/api/main/module.xml (hunk 2 above)
Comment (jamezp) [reply servlet-feature-pack/src/main/resources/modules/system/layers/base/javax/json/bind/api/main/module.xml]: I should have noted too that this change will require the change to use JSON-P 1.1 too as that is the dependency required for JSON-B 1.0. We could just add the two commits from #11370 to this PR too.

File: pom.xml (hunk 3 above)
Comment (asoldano) [reply pom.xml]: As I've written above, I'm going to update the PR here to use 3.6.0.Final before the PR is actually merged.

File: servlet-feature-pack/src/main/resources/modules/system/layers/base/javax/json/bind/api/main/module.xml (hunk 2 above)
Comment (asoldano) [reply servlet-feature-pack/src/main/resources/modules/system/layers/base/javax/json/bind/api/main/module.xml]: I would be fine with merging #11370 first and then rebasing on top of that. What's preventing #11370 from being merged currently?

File: galleon-pack/pom.xml (hunk 1 above)
Comment (asoldano) [reply galleon-pack/pom.xml]: Removed, thanks

**Comment by jamezp** [comment]
@asoldano #11370 has been merged. This will require a rebase.

**Comment by asoldano** [comment]
@jamezp @kabir , rebased

File: jaxrs/src/main/java/org/jboss/as/jaxrs/deployment/JaxrsIntegrationProcessor.java [hunk 4]
```diff
@@ -175,6 +179,15 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
             setContextParameter(webdata, ResteasyContextParameters.RESTEASY_UNWRAPPED_EXCEPTIONS, "javax.ejb.EJBException");
         }
 
+        if (findContextParam(webdata, ResteasyContextParameters.RESTEASY_PREFER_JACKSON_OVER_JSONB) == null) {
+            final String prop = System.getProperty(ResteasyContextParameters.RESTEASY_PREFER_JACKSON_OVER_JSONB, null);
```
Comment (jamezp) [inline jaxrs/src/main/java/org/jboss/as/jaxrs/deployment/JaxrsIntegrationProcessor.java]: This should probably use `WildFlySecurityManager.setPropertyPrivileged(ResteasyContextParameters.RESTEASY_PREFER_JACKSON_OVER_JSONB, null)`

**Review by jamezp — APPROVED** [review]
This looks good other than the minor comment I made.

File: jaxrs/src/main/java/org/jboss/as/jaxrs/deployment/JaxrsIntegrationProcessor.java (hunk 4 above)
Comment (asoldano) [reply jaxrs/src/main/java/org/jboss/as/jaxrs/deployment/JaxrsIntegrationProcessor.java]: Fixed, thanks

**Review by marekkopecky — APPROVED** [review]
I'm OK to merge this PR, I add more details to upgrade jira https://issues.jboss.org/browse/WFLY-10632

**Comment by asoldano** [comment]
@jamezp @kabir , what's still preventing this from being merged?

**Pull request merged**